loadgen/loadgen.cc

```diff
@@ -1196,6 +1075,307 @@ struct LogOutputs {
+                AsyncDetail& detail) {
+    detail("Starting BinarySearchForPerformanceConfig by FindPeakPerformance:");
+    detail("BinarySearchToFindPeakPerformance: Starting with [" + l_field +
+           ", " + r_field + ")");
+  });
+
+  std::string msg;
+
+  while (true) {
+    LogDetail([l_field = ToStringPerformanceField<scenario>(l_bound),
+               r_field = ToStringPerformanceField<scenario>(u_bound)](
+                  AsyncDetail& detail) {
+      detail("BinarySearchToFindPeakPerformance: Searching fields using [" +
+             l_field + ", " + r_field + ")");
+    });
+
+    if (IsFinished<scenario>(l_bound, u_bound)) {
+      break;
+    }
+
+    const TestSettingsInternal search_settings =
+        MidOfBoundaries<scenario>(l_bound, u_bound);
+    LogDetail([field = ToStringPerformanceField<scenario>(search_settings)](
+                  AsyncDetail& detail) {
+      detail(
+          "BinarySearchToFindPeakPerformance: The mid point value of bounds: " +
+          field);
+    });
+
+    PerformanceResult pr(IssueQueries<scenario, TestMode::PerformanceOnly>(
+        sut, search_settings, performance_set, sequence_gen));
+    PerformanceSummary perf_summary{sut->Name(), u_bound, std::move(pr)};
+    if (perf_summary.PerfConstraintsMet(&msg)) {
+      SetPerformanceField<scenario>(l_bound, search_settings);
+    } else {
+      SetPerformanceField<scenario>(u_bound, search_settings);
+    }
+  }
+
+  qsl->UnloadSamplesFromRam(performance_set.set);
+
+  return l_bound;
+}
+
+/// \brief Runs the performance mode, templated by scenario.
+template <TestScenario scenario>
+void RunPerformanceMode(SystemUnderTest* sut, QuerySampleLibrary* qsl,
+                        const TestSettingsInternal& settings,
+                        const std::vector<LoadableSampleSet>& loadable_sets,
+                        SequenceGen* sequence_gen) {
+  LogDetail([](AsyncDetail& detail) { detail("Starting performance mode:"); });
+
+  // Use first loadable set as the performance set.
+  const LoadableSampleSet& performance_set = loadable_sets.front();
+  LoadSamplesToRam(qsl, performance_set.set);
+
+  PerformanceResult pr(IssueQueries<scenario, TestMode::PerformanceOnly>(
+      sut, settings, performance_set, sequence_gen));
+
+  sut->ReportLatencyResults(pr.latencies);
+
+  LogSummary(
+      [perf_summary = PerformanceSummary{sut->Name(), settings, std::move(pr)}](
+          AsyncSummary& summary) mutable { perf_summary.Log(summary); });
+
+  qsl->UnloadSamplesFromRam(performance_set.set);
+}
+
+/// \brief Runs the binary search mode, templated by scenario.
+template <TestScenario scenario>
+void FindPeakPerformanceMode(SystemUnderTest* sut, QuerySampleLibrary* qsl,
+                             const TestSettingsInternal& settings,
+                             const std::vector<LoadableSampleSet>& _,
+                             SequenceGen* sequence_gen) {
+  using namespace find_peak_performance;
```

Comment:
Can you find a way to avoid this using statement? It's not clear which identifiers might be overloaded in this function.

Maybe by explicitly writing find_peak_performance:: or by calling a helper in the find_peak_performance namespace?